Classic McEliece Testspecification (#187)
sehlen-bsi/botan-docs: docs/testspec/src/12_pubkey_kem.rst

   |                        | * :srcref:`src/tests/data/pubkey/cmce_kat_hashed.vec`                   |
   +------------------------+-------------------------------------------------------------------------+
   | **Expected Output:**   | Above described test vector files contain expected values for:          |
   |                        |                                                                         |
   |                        | * Classic McEliece Public Key                                           |
   |                        | * Classic McEliece Private Key                                          |
   |                        | * Ciphertext                                                            |
   |                        | * Shared Secret                                                         |
   +------------------------+-------------------------------------------------------------------------+
   | **Steps:**             | For each KAT vector:                                                    |
   |                        |                                                                         |

>    |                        | #. Seed a AES-256-CTR-DRBG with the specified RNG seed                  |

Reviewer: ```suggestion
   |                        | #. Seed an AES-256-CTR-DRBG with the specified RNG seed.                  |
```

   |                        |                                                                         |
   |                        | #. Use the seeded RNG to generate a Classic McEliece key pair and       |

>    |                        |    compare it to the expected public and private key in the test        |

Reviewer: ```suggestion
   |                        |    compare its private key to the expected public and private key in the test        |
```

Author: Both keys are compared with the KAT data.

>    |                        |    vector. This uses the key encoding as implemented in the reference   |
>    |                        |    implementation and hashes the result using SHAKE-256(512) to save    |

Reviewer: ```suggestion
   |                        |    vector. This vector contains the key encoding as implemented in the reference   |
   |                        |    implementation hashed using SHAKE-256(512) to save    |
```

   |                        |    disk space in the KAT vectors.                                       |
   |                        |                                                                         |
   |                        | #. Check that the expected algorithm properties of the generated key    |

>    |                        |    match the generic expectations for KEMs (supports key encapsulation, |

Reviewer: How do we check this? By just executing them once? Is this still only true for the SK? In that case, clarify

Author: The table **PKENC-CMCE-1** describes the generic KEM test case that is the same for all KEMs. I, therefore, more or less copy-pasted the test specification. I think it's ok to leave this part as it is for now. The properties are tested `For each KAT vector` as stated above. How exactly the specific properties are tested is too much detail, in my opinion. The only thing I'll change is writing in the plural `...of the generated keys` to include sk and pk.

   |                        |    reports a key strength in a reasonable interval, etc.).              |
   |                        |                                                                         |

>    |                        | #. Extract the public key from the just generated key pair and compare  |

Reviewer: hashed here as well?

   |                        |    it to the expected value in the test vector.                         |
   |                        |                                                                         |

>    |                        | #. Encode both public and private key, and decode them again.           |

Reviewer: How is that different to the prior step? Just to check en/decoding consistency?

Author: Yeah. Some algorithms apply DER encoding/decoding. Just checks that: 
`decoding(encoding(key)) == key`

   |                        |                                                                         |
   |                        | #. Encapsulate a secret with the just-generated public key after the    |
   |                        |    encode/decode roundtrip (using the same RNG) and compare the         |
   |                        |    resulting shared secret and ciphertext to expected values in the     |

>    |                        |    test vector. Again, the ciphertext is hashed to save disk space.     |

Reviewer: I thought everything except the ciphertext is hashed?

Author: Thanks! It was a copy-paste error ;)

   |                        |                                                                         |
   |                        | #. Decapsulate the just-calculated ciphertext with the private key from |
   |                        |    the encode/decode roundtrip and ensure that the resulting shared     |
   |                        |    secret is equal to the expected value from the test vector.          |
   +------------------------+-------------------------------------------------------------------------+

.. table::
   :class: longtable
   :widths: 20 80

   +------------------------+-------------------------------------------------------------------------+
   | **Test Case No.:**     | PKENC-CMCE-2                                                            |
   +========================+=========================================================================+
   | **Type:**              | Negative Test/Known Answer Tests                                        |
   +------------------------+-------------------------------------------------------------------------+
   | **Description:**       | For a plaintext confirmation (pc) and a non-pc instance:                |

>    |                        | Generate a Classic McEliece keypair using a KAT seed, use the private   |

Reviewer: any seed?

Also: Should we describe which KAT instances are tested?

   +------------------------+-------------------------------------------------------------------------+
   | **Preconditions:**     | None                                                                    |
   +------------------------+-------------------------------------------------------------------------+
   | **Input Values:**      | KAT seed, invalid ciphertext, corresponding invalid shared secret       |
   +------------------------+-------------------------------------------------------------------------+
   | **Expected Output:**   | The correct invalid shared secret                                       |
   +------------------------+-------------------------------------------------------------------------+
   | **Steps:**             | For each available algorithm parameterization:                          |
   |                        |                                                                         |
   |                        | #. Generate a key pair using the KAT seed with AES-256-CTR-DRBG         |
   |                        |                                                                         |

>    |                        | #. Decapsulate the invalid ciphertext (input value).                    |

Reviewer: How do we get invalid ciphertexts? Are they provided by the KAT vector? If so, how did we make it invalid?

Author: > How do we get invalid ciphertexts?

They were generated using fabian.exe ;) I used the valid ciphertext and flipped a single bit. I did this once with the first bit and once with the last bit (to cover a case where only $C_1$ of plaintext confirmation is wrong). I described it briefly in the text above. But you are right, I'll describe it here.
